File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
from collections import deque
def solution(begin, target, words):
    if target not in words:  # 애초에 words리스트에 target값이 없다면 return 0
        return 0
    answer = 0
    
    queue = deque()
    queue.append([begin,0])
    
    while queue:
        
        begin,cnt = queue.popleft()
        if begin == target:
            return cnt
        for i in words:
            count = 0
            for j in range(len(begin)):
                if i[j] != begin[j]:
                    count+=1
            if count == 1:
                queue.append([i,cnt+1])
                    
                
                
    
    return 0

File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.java
import java.util.*;

class dataPoint{
    String str;
    int count;
    dataPoint(String str, int count){
        this.str = str;
        this.count = count;
    }
}
class Solution {
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        for(String s : words){
            if(s.equals(target)){
                answer ++;
            }
        }
        if(answer != 1){
            return 0;
        }
        answer = 0;
        Deque<dataPoint> data = new LinkedList<>();
        data.add(new dataPoint(begin,0));
        
        while(!data.isEmpty()){
            dataPoint temp = data.poll();
            String checkStr = temp.str;
            int checkCount = temp.count;
            if(checkStr.equals(target)){
                return checkCount;
            }            
            for(String s : words){
                int cnt = 0;
                for(int i=0; i<s.length(); i++){
                    if(checkStr.charAt(i) != s.charAt(i)){
                        cnt++;
                    }
                }
                if(cnt == 1){
                    data.offer(new dataPoint(s,checkCount+1));
                }
            }
        }
        return answer;
    }
}